enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,254 @@
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+
+#### Disaster recovery support
+
+In order to support disaster recovery kubelet, kube-controller-manager and kube-scheduler should be able to communicate with kube-api even when all CHNSO pods are shutdown.
+
+The following steps will ensure that kube-api is available post disaster recovery:
+- Dnsmasq service and Keepalived static pods will start running on master nodes
+- API-VIP address will be assigned to one of the master nodes by Keepalived
+- Dnsmasq should resolve api-int hostname to API-VIP address
+
+The above steps should be applied also for cluster shutdown and resume use-case.
+
+### cluster-hosted-net-services operator
+
+The `cluster-hosted-net-services-operator` is a new operator whose operand is a controller for the cluster hosted DNS, LB and VIP services.
+
+### Standard SLO Behavior
+
+The CHNSO is expected to follow the standard behavior of SLO, additionally, since the CHNSO is applicable only to clusters running on on-prem ( 'baremetal', 'openstack', 'vsphere' or 'ovirt') infrastructure,
+It should support also the expected behaviors for "not in use" SLOs similar to the [CBO operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#standard-slo-behaviors).
+
+### cluster-hosted-net-services-operator details
+
+The CHNSO should:
+
+- Be a new `openshift/cluster-hosted-net-services-operator` project.
+- Publish an image called `cluster-hosted-net-services-operator`.
+- Add a new `ClusterOperator` with an additional
+  `Disabled` status for non on-prem platforms.
+- Implement a controller reconciling a singleton instance representing the desired cluster hosted LB/DNS/VIP configuration.
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource is a platform
+  type other than openstack, baremetal, ovirt or vsphere.
+
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhostednetservices.operator.io/v1alpha1
+kind: ClusterHostedNetServicesConfig
+metadata:
+  name: cluster
+spec:
+  dns: 
+    nodesResolution: Enable
+    appsResolution: Enable    
+    apiintIpAddress: 192.168.111.5 
+  ingressHa: Enable
+status:
+  apiServerInternalIpOwner: master-0
```

Comment:
who will update this? Can keepalived to that? Will there be another process? Or the operator? How does it know the keepalived state.

Reply:
I thought that Keepalived should do that, but I'm not sure if a static pod can (permission) update the CR status - need to check that.
Do you know any code that covered similar logic?

Reply:
keepalived knows about kube apis? 

A static pod has no service account. So you have to "organize" some kubeconfig from somewhere.

Reply:
Currently, we run a sidecar container in the Keepalived pod that retrieves nodes details via kube-api (it uses host's kubelet kubeconfig /var/lib/kubelet/kubeconfig) .

Any ideas on how to "organize" such kubeconfig ? 

How about Keepalived sidecar container in each node annotates his status in the relevant node object and the operator should use this info to update the status?

Reply:
> How about Keepalived sidecar container in each node annotates his status in the relevant node object and the operator should use this info to update the status?

could work, yes

> Any ideas on how to "organize" such kubeconfig ?

kubelet has one, SDN has one. I guess they get it from installer.

Reply:
Good, I assume we'll go with node annotation approach